native/src/main/java/tech/sourced/babelfish/JsonASTVisitor.java

```diff
@@ -131,51 +130,52 @@ private void serializeComments(IASTNode node) throws IOException {
     // We need to call this on each visitor instead of just retuning
     // PROCESS_SKIP and let the base clase do it automatically because
     // we need to start the child JSON array
-    public void visitChilds(IASTNode node) throws IOException {
+    private void visitChildren(IASTNode node) throws IOException {
         IASTNode[] children = node.getChildren();
+        if (children == null || children.length == 0)
+            return;
 
-        if (children != null && children.length > 0) {
-            json.writeFieldName("childs");
-            json.writeStartArray();
-            try {
-                for (IASTNode child : children) {
-                    child.accept(this);
-                }
-            } finally {
-                json.writeEndArray();
+        // Load the children into a hashtable, then write children with the same role
```

Comment:
This works well, but the consequence is that we don't know if a particular field is an array or not.

It should be possible to reflect on `node`, find the field with the same name as `Role` and check if it's a list or not.

Reply:
The field with the same name as Role will be a string field with a description. For example, if you check [this](https://help.eclipse.org/photon/index.jsp?topic=%2Forg.eclipse.cdt.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fcdt%2Fcore%2Fdom%2Fast%2Fcpp%2FICPPASTFunctionDeclarator.html) node, you'll see that every function returning a child or arrays of children has a field (data member) of type [`AstNodeProperty`](https://help.eclipse.org/photon/index.jsp?topic=%2Forg.eclipse.cdt.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fcdt%2Fcore%2Fdom%2Fast%2FASTNodeProperty.html) but if you check those you'll see that they are only a wrapped string (without any indication of the real children being an array or not).

I'll evaluate today another alternative: introspect the node for all methods called "getSomething", then examine the return value and if it's either an instance of `IASTNode` or `IASTNode[]` put the children into a `Something` property (in an array or not). This introspection, if works (I'll have to check against the current approach calling `getChildren` to see if something is missing), then will be cached by node type to speed up things. Maybe I could also introspect all the node types from the library and generate a static mapping, we'll see.

Reply:
PS: I've this working, trough it's pretty horrible. I'll make it pretty before the next commit (and work on catching after that).

Reply:
Done. Took me a while to fix a bug related with some nodes not being visited and the JSON output getting mangled (because some global options needed to be set). Please check the output now, I prefixed the properties holding childen "Prop_" to make them easier to identify (will probably remove that before merge).

The catching/optimization I spoke above is the next step and it'll be certainly needed since now `huge.cpp` randomly fails to generate the fixtures because of a `i/o timeout`  ¯\(°_o)/¯